### PR TITLE
KAFKA-12950: Replace EasyMock and PowerMock with Mockito for KafkaStreamsTest

### PR DESCRIPTION
**Pull request by divijvaidya**
This PR build on top of https://github.com/apache/kafka/pull/11017. I have added the previous author's comment in this PR for attribution. I have also addressed the pending comments from @chia7712 in this PR.

Notes to help the reviewer:
1. Mockito has `mockStatic` method which is equivalent to PowerMock's method.
2. When we run the tests using `@RunWith(MockitoJUnitRunner.StrictStubs.class)` Mockito performs a verify() for all stubs that are mentioned, hence, there is no need to explicitly verify the stubs (unless you want to verify the number of times etc.). Note that this does not work for static mocks.

## Testing
Validated that test is run successfully as part of `./gradlew streams:unitTest` 

### Before
Screenshot 2022-08-02 at 19 02 42

### After
Screenshot 2022-08-02 at 19 02 49

### CODE REVIEW

**Comment by divijvaidya** [comment]
@ijuma Please review when you get a chance. This would help us take a step forward in getting rid of easymock/powermock dependency.

**Comment by divijvaidya** [comment]
@cadonna requesting your review on this PR related to streams.

**Comment by divijvaidya** [comment]
@cadonna please review when you get a chance. You can also refer to previous PR https://github.com/apache/kafka/pull/12465 and note that the comments provided there have been addressed.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -252,10 +252,15 @@ State setState(final State newState) {
...
-            return state.isAlive();
+            return isStateAlive();
         }
     }
 
+    // Ensure Mockito can stub method for KafkaStreamTest.
+    public boolean isStateAlive() {
+        return state.isAlive();
+    }
+
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Why is this needed?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 2]
```diff
@@ -990,7 +990,12 @@ private StreamThread createAndAddStreamThread(final long cacheSizePerThread, fin
         return streamThread;
     }
 
-    static Metrics getMetrics(final StreamsConfig config, final Time time, final String clientId) {
+    // Ensure Mockito stub construct with capture argument for KafkaStreamsTest.
+    public static Metrics createMetrics(final MetricConfig metricConfig, final List<MetricsReporter> reporters, final Time time, final MetricsContext metricsContext) {
+        return new Metrics(metricConfig, reporters, time, metricsContext);
+    }
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: You cannot add a public method call to a public class without a KIP. IMO, it is also not good to add methods just for tests to a class. It often points to code-smell.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 3]
```diff
@@ -192,154 +209,141 @@ public void before() throws Exception {
...
-            anyObject(MetricsContext.class)
-        ).andAnswer(() -> {
+        kafkaStreamsMockedStatic = mockStatic(KafkaStreams.class, withSettings()
+                .defaultAnswer(InvocationOnMock::callRealMethod));
+        kafkaStreamsMockedStatic.when(() -> KafkaStreams.createMetrics(
+                any(MetricConfig.class),
+                metricsReportersCapture.capture(),
+                any(Time.class),
+                any(MetricsContext.class)
+        )).thenAnswer(invocation -> {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: You might be able to mock the `Metrics.class` with `mockConstruction()` (see https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#mocked_construction). Some instructions can also be found at https://www.davidvlijmincx.com/posts/mockito_mock_constructor/.
In such a way you would avoid the change of the public API in `KafkaStreams`.

**Review by cadonna — COMMENTED** [review]
@divijvaidya Thanks for the PR!

I haven't done a complete pass but I found an issue that you should fix right away.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 3 above)
Comment (divijvaidya) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Thank you for the suggestion. It took a while but figured out a way to do what we wanted to in the test without having to change any public API.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 2 above)
Comment (divijvaidya) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I have changed this in the latest commit.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (divijvaidya) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: This has been removed in the latest commit.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 4]
```diff
@@ -1096,7 +1096,7 @@ private Optional<String> removeStreamThread(final long timeoutMs) throws Timeout
                 // make a copy of threads to avoid holding lock
                 for (final StreamThread streamThread : new ArrayList<>(threads)) {
                     final boolean callingThreadIsNotCurrentStreamThread = !streamThread.getName().equals(Thread.currentThread().getName());
-                    if (streamThread.isAlive() && (callingThreadIsNotCurrentStreamThread || getNumLiveStreamThreads() == 1)) {
```
Comment (divijvaidya) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1100]: Note that `streamThread.isAlive()` is a `native` method and Mockito currently does not support mocking of `native` methods. 

I have made this change in logic to remove the check because even if the thread is not running, we should remove it from data structures such as `threads` and perform the clean up of the Java object. Hence, the check is not really necessary.

@cadonna let me know if you have better ideas here.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 5]
```diff
@@ -1146,6 +1146,10 @@ private long advanceNowAndComputeLatency() {
     public void shutdown() {
         log.info("Informed to shut down");
         final State oldState = setState(State.PENDING_SHUTDOWN);
+
+        // If the thread is already shutting down, this is a No-op.
+        if (oldState == State.PENDING_SHUTDOWN || oldState == State.DEAD) return;
```
Comment (divijvaidya) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I would like to piggy back this change with this commit. Happy to open a new PR too if you think that is required.

**Comment by divijvaidya** [comment]
@cadonna this is ready for your review whenever you get a chance.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 5 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I think the condition is never true. Method `setState()` will return `null` if the current state is either `PENDING_SHUTDOWN` (https://github.com/divijvaidya/kafka/blob/1780a0c4676f212b1271688248383420e754718f/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L224)  or `DEAD` (https://github.com/divijvaidya/kafka/blob/1780a0c4676f212b1271688248383420e754718f/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L230).

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Even if you check for `null`, `shutdown()` is also without that check a no-op when the thread is already shutting down.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 6]
```diff
@@ -1096,7 +1096,7 @@ private Optional<String> removeStreamThread(final long timeoutMs) throws Timeout
...
+                    if (callingThreadIsNotCurrentStreamThread || getNumLiveStreamThreads() == 1) {
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: @wcarlson5 Could you please help me understanding the implications of removing this check? 
My take is that: 

Without the check:
- If the stream thread is not alive and its state is not `DEAD`, it will not be removed from the list of threads. 
- If the stream thread is not alive and its state is `DEAD`, it will be removed from the list of threads.
- If the stream thread is alive and its state is not `DEAD`, it will be removed from the list of threads once it reaches `DEAD` -- here, in `getNumLiveStreamThreads()`, or in `getNextThreadIndex()`. 
- If the stream thread is alive and its state is `DEAD`, it will be removed from the list of threads.

With the check:
- If the stream thread is not alive and its state is not `DEAD`, it will not be removed from the list of threads. 
- If the stream thread is not alive and its state is `DEAD`, it will be removed from the list of threads in `getNumLiveStreamThreads()` or in `getNextThreadIndex()` but not here.
- If the stream thread is alive and its state is not `DEAD`, it will be removed from the list of threads once it reaches `DEAD` -- here, in `getNumLiveStreamThreads()`, or in `getNextThreadIndex()`. 
- If the stream thread is alive and its state is `DEAD`, it will be removed from the list of threads.

So it seems the check does not change much except that it delays the removal from the list if the stream thread is not alive and its state is `DEAD`.

Did I miss a case?

For context, this check blocks the migration to Mockito since Mockito cannot stub native methods and `isAlive()` is a native method.

Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: So this check is important and I don't think we can just remove it.

The important thing this is doing is preventing a thread from being removed that is already dead. If a thread is removed but not finished cleaning up it is still in the thread list. If we then call remove threads and that dead thread is picked then no thread will really be removed so that call is broken. Even worse though the cache size of all the other threads are changed to some value that is not correct.

Instead of just removing the check you can maybe change to do something like 
`thread.state() == StreamThread.State.DEAD`

Maybe you can add a helper method to check that instead of using the native methods if w really need to.

On a side note it is concerning to me that we can't mock native methods anymore that does not seem a good thing. Why are we removing this ablity?

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: @wcarlson5 Thanks for the info! Really useful! 

> Instead of just removing the check you can maybe change to do something like
> thread.state() == StreamThread.State.DEAD

I guess you mean `thread.state() != StreamThread.State.DEAD && thread.state() != StreamThread.State.PENDING_SHUTDOWN`
I think we also need to add `thread.state() != StreamThread.State.PENDING_SHUTDOWN` since stream threads in `StreamThread.State.PENDING_SHUTDOWN` are skipped in `getNumLiveStreamThreads()` and the result of that method is used to resize the cache. 

I am not too much concerned about losing the ability to mock native methods. I think it is much more important to use a mocking framework that is actively maintained and does not block Java version upgrades in AK. As far as I understand that is the reason we move away from EasyMock and PowerMock to Mockito.

Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: You can look at the is alive method for state checks that are happening and refactor it into a method that works for Mockito. 

Ah that makes more sense. Thanks for clarifying

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Wouldn't it be possible to just call `streamThread.state().isAlive()`?

Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Maybe, if the mock works then proabably

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: That is not a native method. A native method looks like the following in `java.lang.Thread`:
```
public final native boolean isAlive();
```

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Ah, I think you mean because it is a method on an enum.

Comment (divijvaidya) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Thank you both for the discussion here.

I believe that we cannot use `streamThread.state().isAlive()` here because a thread in `CREATED` state is not considered `isAlive()` by this method which can lead to the following regression from existing behaviour:
1. Inside `addStreamThread()`, thread gets created and added to `threads` list at `createAndAddStreamThread()` but Thread.start() (which transitions the thread state to STARTING) hasn't been called yet (because, let's say current thread is waiting at line 1033 in addStreamThread() method).
2. `removeStreamThread` gets called by another thread.

**Old behaviour:**

`streamThread.isAlive()` will return true and the thread will be removed from the list. When the first thread 

**New behaviour:**

`streamThread.state().isAlive()` will return false and the thread will not be removed from the list, hence `removeStreamThread` will not remove this thread.

I believe that this new behaviour is a change in public interface contract for `removeStreamThread` and hence, perhaps could be a breaking change.

Alternatively, 
the simplest solution is to create a new method which is a wrapper over the native method such as:
```
// Ensure Mockito can stub method for KafkaStreamTest.
    public boolean isStateAlive() {
        return state.isAlive();
    }
```
Is that something that would be acceptable to you folks?

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Since we are talking about the migration from EasyMock to Mockito and not about improving the current behavior, I would go with the simplest solution. 

I think, you meant to write:
```java
public boolean isThreadAlive() {
    return isAlive();
}
```
 
Thank you for proposing that solution!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 7]
```diff
@@ -1081,6 +1081,11 @@ private void addToResetList(final TopicPartition partition, final Set<TopicParti
         partitions.add(partition);
     }
 
+    // This method is added for usage in tests where mocking the underlying native call is not possible.
+    public boolean isThreadAlive() {
```
Comment (divijvaidya) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1086]: Note that adding this public method does not require a KIP because this class is not part of the public API at https://kafka.apache.org/32/javadoc/allclasses-index.html

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 5 above)
Comment (divijvaidya) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Thanks for pointing this out. I missed reading the setState code. I have reverted this change.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 6 above)
Comment (divijvaidya) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I have added this new method. Note that adding this public method does not require a KIP because this class is not part of the public API at https://kafka.apache.org/32/javadoc/allclasses-index.html

This should resolve the concern with this change as it doesn't change the behaviour now.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 4 above)
Comment (divijvaidya) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1100]: This is resolved. See the conversation at https://github.com/apache/kafka/pull/12465#discussion_r984479081

**Comment by divijvaidya** [comment]
Rebased from latest commits in trunk

**Review by wcarlson5 — APPROVED** [review]
I think this is ready to merge

**Review by bbejeck — APPROVED** [review]
LGTM

**Comment by bbejeck** [comment]
Failures unrelated

**Comment by bbejeck** [comment]
Merged #12465 into trunk

**Comment by bbejeck** [comment]
Thanks for the contribution @divijvaidya !

**Comment by dengziming** [comment]
@divijvaidya So can we close #11017 after this is merged?

**Comment by divijvaidya** [comment]
@dengziming yes please. It can be closed.
